src/Nethermind/Nethermind.Serialization.Rlp/TxDecoder.cs

```diff
@@ -2,887 +2,160 @@
 // SPDX-License-Identifier: LGPL-3.0-only
 
 using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
 using Microsoft.Extensions.ObjectPool;
 using Nethermind.Core;
-using Nethermind.Core.Crypto;
-using Nethermind.Core.Extensions;
-using Nethermind.Serialization.Rlp.Eip2930;
+using Nethermind.Serialization.Rlp.MyTxDecoder;
 
-namespace Nethermind.Serialization.Rlp
+namespace Nethermind.Serialization.Rlp;
+
+public sealed class TxDecoder : TxDecoder<Transaction>
 {
-    public sealed class TxDecoder : TxDecoder<Transaction>
-    {
-        public const int MaxDelayedHashTxnSize = 32768;
-        public static TxDecoder Instance = new TxDecoder();
-        public static TxDecoder InstanceWithoutLazyHash = new TxDecoder(false);
-        public static ObjectPool<Transaction> TxObjectPool = new DefaultObjectPool<Transaction>(new Transaction.PoolPolicy(), Environment.ProcessorCount * 4);
+    public static readonly ObjectPool<Transaction> TxObjectPool = new DefaultObjectPool<Transaction>(new Transaction.PoolPolicy(), Environment.ProcessorCount * 4);
 
-        public TxDecoder() : base(true) // Rlp will try to find empty constructor.
-        {
-        }
+    public TxDecoder() : base(true) { } // Rlp will try to find empty constructor.
+    public TxDecoder(bool lazyHash) : base(lazyHash) { }
+    protected override Transaction NewTx() => TxObjectPool.Get();
+}
 
-        public TxDecoder(bool lazyHash) : base(lazyHash)
-        {
-        }
+public sealed class SystemTxDecoder : TxDecoder<SystemTransaction> { }
+public sealed class GeneratedTxDecoder : TxDecoder<GeneratedTransaction> { }
 
-        protected override Transaction NewTx()
-        {
-            return TxObjectPool.Get();
-        }
-    }
-    public class SystemTxDecoder : TxDecoder<SystemTransaction> { }
-    public class GeneratedTxDecoder : TxDecoder<GeneratedTransaction> { }
+public class TxDecoder<T> : IRlpStreamDecoder<T>, IRlpValueDecoder<T> where T : Transaction, new()
+{
+    private readonly Dictionary<TxType, ITxDecoder> _decoders;
 
-    public class TxDecoder<T> :
-        IRlpStreamDecoder<T>,
-        IRlpValueDecoder<T>
-        where T : Transaction, new()
+    public TxDecoder(bool lazyHash = true)
     {
-        private readonly AccessListDecoder _accessListDecoder = new();
-        private readonly bool _lazyHash;
-
-        protected TxDecoder(bool lazyHash = true)
-        {
-            _lazyHash = lazyHash;
-        }
+        _decoders = new() {
+            { TxType.Legacy, new LegacyTxDecoder(lazyHash, NewTx) },
+            { TxType.AccessList, new AccessListTxDecoder(lazyHash, NewTx) },
+            { TxType.EIP1559, new EIP1559TxDecoder(lazyHash, NewTx) },
+            { TxType.Blob, new BlobTxDecoder(lazyHash, NewTx) },
+            { TxType.DepositTx, new OptimismTxDecoder(lazyHash, NewTx) }
+        };
+    }
```

Comment:
Should this be static?

Reply:
You mean the decoder list? If so, have this `lazyHash` property that we need to thread to all decoders. If we make them static then we would need to have a copy with a different `lazyHash` value, and it would be hard for the methods to decide which one to pick. I think a possible approach would be the following:

- Keep `_decoders` a property
- Make two static instances of `TxDecoder<T>` where one has `lazyHash = true`, and the other `lazyHash = false`.
- Make the constructor for `TxDecoder<T>` private/protected, thus making it esentially a singleton.

Reply:
Implemented the proposed approach and remove the `lazyHash` parameter.